### PR TITLE
383 notification service (#736)

### PR DESCRIPTION
**Pull request by AlyssonMascarenhas**
## Description

Create a notification service

## Proposed Changes

- Create a notification service

## How to Test

Implements the notification service

## Screenshots

Include screenshots or GIFs that illustrate the proposed changes. This can be especially useful for user interface changes.

## View Storybook
https://62eab350a45bdb0a5818520e-bfcpynydxb.chromatic.com/?path=/story/ion-feedback-notification--service

## Compliance

- [x] I have verified that this change complies with our code and contribution policies.
- [x] I have verified that this change does not cause regressions and does not affect other parts of the code.

### CODE REVIEW

File: projects/ion/src/lib/notification/service/notification.service.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    notification.changeDetectorRef.detectChanges();
+
+    this.notificationContainerComponentRef.instance.addNotification(
+      notification
+    );
+
+    this.notificationContainerComponentRef.changeDetectorRef.detectChanges();
+
+    return this.componentSubscriber.next();
+  }
```
Comment (Deeved) [inline projects/ion/src/lib/notification/service/notification.service.ts:150]: the success, information, warning and error functions are practically the same. I believe you can reduce code repetition by creating, for example, a generic function.

File: projects/ion/src/lib/notification/service/notification.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  it('should create a notification', () => {
+    notificationService.success('Teste', 'Teste');
+    expect(notificationService).toBeTruthy();
+    expect(screen.getByTestId('ion-notification')).toBeTruthy();
+  });
+
+  it('should create a sucess notification', () => {
+    notificationService.success('Teste', 'Teste');
+    const iconType = screen.getAllByTestId('notification-icon');
+    expect(iconType[1].classList).toContain('success-icon');
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.spec.ts:58]: You can use tohaveclass to this case, is more simple and readability

File: projects/ion/src/lib/notification/service/notification.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  });
+
+  it('should create a info notification', () => {
+    notificationService.info('Teste', 'Teste');
+    const iconType = screen.getAllByTestId('notification-icon');
+    expect(iconType[2].classList).toContain('info-icon');
+  });
+
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.spec.ts:66]: You can improve this cases of type with it.each(['warning', 'success'...])

File: projects/ion/src/lib/notification/service/notification.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  it('should create a warning notification', () => {
+    notificationService.warning('Teste', 'Teste');
+    const iconType = screen.getAllByTestId('notification-icon');
+    expect(iconType[3].classList).toContain('warning-icon');
+  });
+
+  it('should create a negative notification', () => {
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.spec.ts:73]: I believe that you can make a test opening multiples notifications and remove last, first and second.

File: projects/ion/src/lib/notification/service/notification.service.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+  ): void {
+    notification.title = title;
+    notification.message = message;
+    notification.type = type;
+
+    if (options) {
+      Object.keys(options).forEach((prop) => {
+        notification[prop] = options[prop];
+      });
+    }
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.ts:196]:
```suggestion
    if (options) {
      Object.keys(options).forEach((prop) => {
        notification[prop] = options[prop];
      });
    }
```

Do you this to copy object? Case yes, try this.

```
if (options) {
  notification = { ...notification, ...options };
}
```

**Review by iurynogueira — CHANGES_REQUESTED** [review]
Thanks guys, we was need this a lot of time.

spin-chair

File: projects/ion/src/lib/notification/service/notification.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+import { fireEvent, screen } from '@testing-library/angular';
+
+const NOTIFICATION_ICONS = {
+  success: 'success-icon',
+  info: 'info-icon',
+  warning: 'warning-icon',
+  negative: 'negative-icon',
+};
+
+const NOTIFICATION_TYPES = ['success', 'info', 'warning', 'negative'];
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.spec.ts:16]: This is equals to Object.keys(NOTIFICATION_ICONS)

File: projects/ion/src/lib/notification/service/notification.service.spec.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+  it('should remove a notification', () => {
+    notificationService.success('Custom', 'Custom', { fixed: true });
+    const removeNotification = screen.getByTestId('btn-remove');
+    fireEvent.click(removeNotification);
+    const elements = document.getElementsByTagName('ion-notification');
+    expect(elements).toHaveLength(0);
+  });
+
+  it('should remove a notification', () => {
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.spec.ts:58]:
```suggestion
  it('should emit event when a custom notification is closed', () => {
```

File: projects/ion/src/lib/notification/service/notification.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+    notificationService = TestBed.get(IonNotificationService);
+  });
+
+  it.each(NOTIFICATION_TYPES)('should create %s notification', async (type) => {
+    NOTIFICATIONS_CALLS[type]();
+    const iconType = screen.getAllByTestId('notification-icon');
+    expect(iconType[currentIndex]).toHaveClass(NOTIFICATION_ICONS[type]);
+    currentIndex += 1;
+  });
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.spec.ts:114]: nice

File: projects/ion/src/lib/notification/service/notification.service.ts [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    if (!this.notificationContainerComponentRef)
+      this.createNotificationContainer();
+
+    const notification = this.createNotificationInstance();
+
+    this.configNotification(notification.instance, title, message, options);
+
+    this.instanceNotification(notification);
+
+    closeEventCall && this.addCloseEventEmitter(notification, closeEventCall);
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.ts:53]: https://github.com/Brisanet/typescript/blob/main/docs/Style%20Guide.md#control-statements

```
// bad
!isRunning && startRunning();

// good
if (!isRunning) {
  startRunning();
}
```

File: projects/ion/src/lib/notification/service/notification.service.ts (hunk 1 above)
Comment (iurynogueira) [reply projects/ion/src/lib/notification/service/notification.service.ts:126]: I waiting this to approve this awesome PR :)

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: projects/ion/src/lib/notification/service/notification.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+      'Custom',
+      { fixed: true },
+      closeEvent
+    );
+    const removeNotification = screen.getByTestId('btn-remove');
+    fireEvent.click(removeNotification);
+    expect(closeEvent).toHaveBeenCalledTimes(1);
+  });
+
+  it('should create a notification', () => {
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.spec.ts:71]: Create a test case that check title and subtitle

File: projects/ion/src/lib/notification/service/notification.service.ts [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+      message,
+      options,
+      NOTIFICATION_TYPES.warning
+    );
+
+    this.instanceNotification(notification);
+
+    if (closeEventCall) {
+      this.addCloseEventEmitter(notification, closeEventCall);
+    }
```
Comment (iurynogueira) [inline projects/ion/src/lib/notification/service/notification.service.ts:107]: I believe that you can create a method called showNotification and put this code and call method in all methods by type. (sucess, warning) like this:

```suggestion
    this.showNotification(title, message, options, closeEventCall, NOTIFICATION_TYPES.warning)
```

**Review by iurynogueira — CHANGES_REQUESTED** [review]

**Pull request merged**